common/persistence/cassandra/shard_store.go

```diff
@@ -80,55 +80,60 @@ func NewShardStore(
 	}
 }
 
-func (d *ShardStore) CreateShard(
-	request *p.InternalCreateShardRequest,
-) error {
-	query := d.Session.Query(templateCreateShardQuery,
+func (d *ShardStore) GetOrCreateShard(
+	request *p.InternalGetOrCreateShardRequest,
+) (*p.InternalGetOrCreateShardResponse, error) {
+	query := d.Session.Query(templateGetShardQuery,
 		request.ShardID,
 		rowTypeShard,
 		rowTypeShardNamespaceID,
 		rowTypeShardWorkflowID,
 		rowTypeShardRunID,
 		defaultVisibilityTimestamp,
-		rowTypeShardTaskID,
-		request.ShardInfo.Data,
-		request.ShardInfo.EncodingType.String(),
-		request.RangeID)
+		rowTypeShardTaskID)
 
-	previous := make(map[string]interface{})
-	applied, err := query.MapScanCAS(previous)
-	if err != nil {
-		return gocql.ConvertError("CreateShard", err)
+	var data []byte
+	var encoding string
+	err := query.Scan(&data, &encoding)
+	if err == nil {
+		return &p.InternalGetOrCreateShardResponse{
+			ShardInfo: p.NewDataBlob(data, encoding),
+		}, nil
+	} else if !gocql.IsNotFoundError(err) || request.CreateShardInfo == nil {
+		return nil, gocql.ConvertError("GetOrCreateShard", err)
 	}
 
-	if !applied {
-		return &p.ShardAlreadyExistError{
-			Msg: fmt.Sprintf("Shard already exists in executions table.  ShardId: %v.", request.ShardID),
-		}
+	// shard was not found and we should create it
+	rangeID, shardInfo, err := request.CreateShardInfo()
+	if err != nil {
+		return nil, err
 	}
-	return nil
-}
 
-func (d *ShardStore) GetShard(
-	request *p.InternalGetShardRequest,
-) (*p.InternalGetShardResponse, error) {
-	shardID := request.ShardID
-	query := d.Session.Query(templateGetShardQuery,
-		shardID,
+	query = d.Session.Query(templateCreateShardQuery,
+		request.ShardID,
 		rowTypeShard,
 		rowTypeShardNamespaceID,
 		rowTypeShardWorkflowID,
 		rowTypeShardRunID,
 		defaultVisibilityTimestamp,
-		rowTypeShardTaskID)
+		rowTypeShardTaskID,
+		shardInfo.Data,
+		shardInfo.EncodingType.String(),
+		rangeID)
 
-	var data []byte
-	var encoding string
-	if err := query.Scan(&data, &encoding); err != nil {
-		return nil, gocql.ConvertError("GetShard", err)
+	previous := make(map[string]interface{})
+	applied, err := query.MapScanCAS(previous)
+	if err != nil {
+		return nil, gocql.ConvertError("GetOrCreateShard", err)
 	}
-
-	return &p.InternalGetShardResponse{ShardInfo: p.NewDataBlob(data, encoding)}, nil
+	if !applied {
+		// conflict, try again
+		request.CreateShardInfo = nil
+		return d.GetOrCreateShard(request)
```

Comment:
I would try to avoid doing this so we don't have the risk of recursively call in a tight loop and risk of stackoverflow. Instead using retry package to wrap the call so it retry on known errors with backoff and max attempt.

Reply:
That's why it sets CreateShardInfo to nil, to prevent a loop even if cassandra returns nonsense.

Retry would be way overkill here: we just want to get exactly once more with no delay. (Also, callers are probably doing their own retry.)

Reply:
o, good i missed that.